index.js

```diff
@@ -1,9 +1,93 @@
 const daggy = require('daggy')
 const Dict = daggy.tagged('data')
+const Option = require('fantasy-options')
+const Either = require('fantasy-eithers')
+const Identity = require('fantasy-identities')
+
 module.exports = Dict
 
+// Helper functions (shamelessly ripoff from https://github.com/component/clone)
+// slight modification to have clones share their prototype
+// [TODO] submit a Pull Request to original project and use it as an external dependency for object cloning ?
+function type (val) {
+  switch (toString.call(val)) {
+    case '[object Date]': return 'date'
+    case '[object RegExp]': return 'regexp'
+    case '[object Arguments]': return 'arguments'
+    case '[object Array]': return 'array'
+    case '[object Error]': return 'error'
+  }
+  if (val === null) return 'null'
+  if (val === undefined) return 'undefined'
+//  if (val !== val) return 'nan'
+  if (val && val.nodeType === 1) return 'element'
+
+  val = val.valueOf
+    ? val.valueOf()
+    : Object.prototype.valueOf.apply(val)
+
+  return typeof val
+}
+
+function clone (obj) {
+  switch (type(obj)) {
+    case 'object':
+
+      // preserve prototype of cloned object
+      copy = Object.create(Object.getPrototypeOf(obj))
+
+      for (var key in obj) {
+        if (Object.prototype.hasOwnProperty.call(obj, key)) {
+          copy[key] = clone(obj[key])
+        }
+      }
+      return copy
+
+    case 'array':
+      var copy = new Array(obj.length)
+      for (var i = 0, l = obj.length; i < l; i++) {
+        copy[i] = clone(obj[i])
+      }
+      return copy
+
+    case 'date':
+      return new Date(obj.getTime())
+
+    default: // string, number, boolean, …
+      return obj
+  }
+}
+
+// Fantasy Land has no interface to test if an object is a fantasy-land value
+//   - isADT should provide compatibility with most used Fantasy-land compatible libs (folktale, Sanctuary, Ramda-fantasy, etc.)
+//   - current implementation tests Maybe and Either Monads and should be conservatively extended to to other usefull monads
```

Comment:
I feel this type of logic should generally not be necessary in working with fantasy-land types.

Firstly, because types are easily recognizable by the methods they have (`.chain()` in case of Monads, `.map()` in case of functors, etc). Though this could of course lead to false positives.

But more fundamentally, because the fantasy-land specification described what kind of types you are going to get, so you shouldn't need to infer this at runtime. Taking the example of `sequence()`: it assumes it's passed a traversable with children that are Applys. Getting something else implies a programatic error of some sort.

The beauty of this is that code relying simply on for instance the `.ap()` method any Apply should have, is that you're automatically compatible with all Apply's following the specification. Supporting specific implementations of some types should not be necessary.